src/main/java/com/yelp/nrtsearch/server/utils/LRUCache.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright 2020 Yelp Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied.
+ * See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package com.yelp.nrtsearch.server.utils;
+
+import java.util.AbstractMap;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+
+/**
+ * Cache to hold the last n elements accessed, optimized for readers. The class is thread safe
+ * through use of a {@link ReentrantReadWriteLock}. Interface functions that preform read operation only
+ * use the read lock. Functions that modify cache membership (add/remove/clear/etc) use the write lock.
+ * Read operations are O(1) and write operations are O(n) if an eviction is required.
+ *
+ * Each cache entry contains the last time it was accessed. This will be either the insertion time, or
+ * the time of the last get operation. The timestamp is declared volatile to ensure atomic writes and
+ * visibility.
+ * @param <K> type of cache key
+ * @param <V> type of cache value
+ */
+public class LRUCache<K, V> implements Map<K, V> {
+    private static class CacheEntry<T> {
+        T value;
+        volatile long lastAccessTimeMs;
+
+        CacheEntry(T value) {
+            this.value = value;
+            lastAccessTimeMs = System.currentTimeMillis();
+        }
+    }
+
+    private final Map<K, CacheEntry<V>> cache = new HashMap<>();
```

Comment:
could we use `LinkedHashMap` here to get the LRU behavior like below ? Ofc you will need to still use the locks for concurrency?

```
public LRUCache(int capacity) {
            CAPACITY = capacity;
            map = new LinkedHashMap<Integer, Integer>(capacity, 0.75f, true){
                protected boolean removeEldestEntry(Map.Entry eldest) {
                    return size() > CAPACITY;
                }
            };
        }
        public int get(int key) {
            return map.getOrDefault(key, -1);
        }
        public void set(int key, int value) {
            map.put(key, value);
        }
```

Reply:
Can also use https://github.com/ben-manes/caffeine

Reply:
I guess it would be worth trying the Guava cache. It will occasionally add some extra work on get calls, which I was trying to avoid, but it does simplify things.